src/Storages/StorageMerge.cpp

```diff
@@ -666,6 +669,44 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
                 processed_stage,
                 max_block_size,
                 UInt32(streams_num));
+
+
+            if (!plan.isInitialized())
+                return {};
+
+            if (auto * source_step_with_filter = dynamic_cast<SourceStepWithFilter*>((plan.getRootNode()->step.get())))
+            {
+                auto row_policy_filter = modified_context->getRowPolicyFilter(database_name, table_name, RowPolicyFilterType::SELECT_FILTER);
+
+                if (row_policy_filter)
+                {
+                    ASTPtr expr = row_policy_filter->expression;
+
+                    auto syntax_result = TreeRewriter(modified_context).analyze(expr, header.getNamesAndTypesList());
+                    auto expression_analyzer = ExpressionAnalyzer{row_policy_filter->expression, syntax_result, modified_context};
+
+                    auto filter_dag_ptr = expression_analyzer.getActionsDAG(true, false);
+                    auto filter_actions = std::make_shared<ExpressionActions>(filter_dag_ptr, ExpressionActionsSettings::fromContext(modified_context, CompileExpressions::yes));
+                    auto required_columns = filter_actions->getRequiredColumns();
+
+                    LOG_TRACE(&Poco::Logger::get("ReadFromMerge::convertinfSourceStream"), "filter_actions_dag: {},<> {}, <> {}",
+                        filter_actions->getActionsDAG().dumpNames(), filter_actions->getActionsDAG().dumpDAG(), filter_actions->getSampleBlock().dumpStructure());
+
+                    auto fa_actions_columns_sorted = filter_actions->getSampleBlock().getNames();
+                    std::sort(fa_actions_columns_sorted.begin(), fa_actions_columns_sorted.end());
+
+                    Names required_columns_sorted = required_columns;
+                    std::sort(required_columns_sorted.begin(), required_columns_sorted.end());
+
+                    Names filter_columns;
+
+                    std::set_difference(fa_actions_columns_sorted.begin(), fa_actions_columns_sorted.end(),
+                        required_columns.begin(), required_columns.end(),
+                        std::inserter(filter_columns, filter_columns.begin()));
+
+                    source_step_with_filter->addFilter(filter_dag_ptr, filter_columns.front());
```

Comment:
Why only with columns from RBAC that are not required columns? Would filter work properly if there are some required columns in expression?

Reply:
Why only 1 filter, what if there are many columns?

Reply:
The 'extra' column here is a complex condition that may contain logical operations, something like
'field1 > x and field2 > y'.
That's why only one filter is needed.
I'll try to simulate collision between the column and what is listed in required column, this is an interesting point.

Reply:
I believe that it might be beneficial to put that info into comment for this piece of code